ir/ir-traversal.h

```diff
@@ -0,0 +1,202 @@
+struct Traverse {
+    template <typename Obj, typename T>
+    static const Obj *modify(const Obj *, Assign<const T *> asgn) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        return asgn.value;
+    }
+
+    template <typename Obj, typename T>
+    static const Obj *modify(Obj *, Assign<const T *> asgn) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        return asgn.value;
+    }
+
+    template <typename Obj, typename T>
+    static Obj *modify(Obj *obj, Assign<T> &&asgn) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        *obj = std::move(asgn.value);
+        return obj;
+    }
+
+    template <typename Obj, typename T>
+    static Obj *modify(Obj *obj, const Assign<T> &asgn) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        *obj = asgn.value;
+        return obj;
+    }
+
+    template <typename Obj, typename Fn>
+    static decltype(auto) modify(Obj *obj, Fn fn) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        return fn(obj);
+    }
+
+    template <typename Obj, typename... Selectors>
+    static Obj *modify(Obj *obj, Index idx, Selectors &&...selectors) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        BUG_CHECK(obj->size() > idx.value, "Index %1% out of bounds of %2%", idx.value, obj);
+        modifyRef((*obj)[idx.value], std::forward<Selectors>(selectors)...);
+        return obj;
+    }
+
+    template <typename Obj, typename To, typename... Selectors>
+    static To *modify(Obj *obj, RTTI::Detail::ToType<To> cast, Selectors &&...selectors) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        auto *casted = cast(obj);
+        BUG_CHECK(casted, "Cast of %1% failed", obj);
+        return modify(casted, std::forward<Selectors>(selectors)...);
+    }
+
+    template <typename Obj, typename T, typename Sub, typename... Selectors>
+    static Obj *modify(Obj *obj, Sub T::*member, Selectors &&...selectors) {
+        static_assert(!std::is_const_v<Obj>, "Cannot modify constant object");
+        modifyRef(obj->*member, std::forward<Selectors>(selectors)...);
+        return obj;
+    }
+
+    template <typename T, typename... Selectors>
+    static void modifyRef(T &ref, Selectors &&...selectors) {
+        if constexpr (std::is_pointer_v<T>) {
+            ref = modify(ref->clone(), std::forward<Selectors>(selectors)...);
+        } else {
+            auto *res = modify(&ref, std::forward<Selectors>(selectors)...);
+            if (&ref != res) {
+                ref = *res;
+            }
+        }
+    }
+};
+
+}  // namespace Detail
+
+/// @brief Given an object @p obj and a series of selector (ending with a modifier), modify the @p
+/// obj's sub object at the selecte path. %This is useful for deeper modification of objects that
```

Comment:
```suggestion
/// obj's sub object at the selected path. This is useful for deeper modification of objects that
```

Reply:
The `%` is there actually for a reason, although it is an annoying one. Otherwise Doxygen picks `IR::This` and adds links to it, which is quite annoying.